File: implementation/scipamato/scipamato-core-web/src/main/java/ch/difty/scipamato/core/web/keyword/KeywordEditTranslationPanel.java
package ch.difty.scipamato.core.web.keyword;

import java.util.Iterator;

import de.agilecoders.wicket.core.markup.html.bootstrap.button.ButtonBehavior;
import de.agilecoders.wicket.extensions.markup.html.bootstrap.confirmation.ConfirmationBehavior;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.ajax.markup.html.AjaxLink;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.Form;
import org.apache.wicket.markup.html.form.TextField;
import org.apache.wicket.markup.repeater.Item;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.StringResourceModel;

import ch.difty.scipamato.core.entity.keyword.KeywordDefinition;
import ch.difty.scipamato.core.entity.keyword.KeywordTranslation;
import ch.difty.scipamato.core.web.common.DefinitionEditTranslationPanel;

@SuppressWarnings({ "SameParameterValue", "WicketForgeJavaIdInspection" })
abstract class KeywordEditTranslationPanel
    extends DefinitionEditTranslationPanel<KeywordDefinition, KeywordTranslation> {

    KeywordEditTranslationPanel(final String id, final IModel<KeywordDefinition> model) {
        super(id, model);
    }

    @Override
    protected void addColumns(final Item<KeywordTranslation> item) {
        item.add(new Label("langCode"));
        item.add(new TextField<String>("name"));
        item.add(newAddLink("addTranslation", item));
        item.add(newRemoveLink("removeTranslation", item));
    }

    private AjaxLink<Void> newAddLink(final String id, final Item<KeywordTranslation> item) {
        final AjaxLink<Void> newLink = new AjaxLink<>(id) {
            private static final long serialVersionUID = 1L;

            @Override
            public void onClick(final AjaxRequestTarget target) {
                final KeywordTranslation currentKt = item.getModelObject();
                final String langCode = currentKt.getLangCode();
                final KeywordTranslation newKt = new KeywordTranslation(null, langCode, null, 1);
                KeywordEditTranslationPanel.this
                    .getModelObject()
                    .getTranslations()
                    .put(langCode, newKt);
                target.add(getForm());
            }
        };
        newLink.add(new ButtonBehavior());
        newLink.setBody(new StringResourceModel("button." + id + LABEL_RESOURCE_TAG));
        return newLink;
    }

    private AjaxLink<Void> newRemoveLink(final String id, final Item<KeywordTranslation> item) {
        final AjaxLink<Void> newLink = new AjaxLink<>(id) {
            private static final long serialVersionUID = 1L;

            @Override
            public void onClick(final AjaxRequestTarget target) {
                final KeywordTranslation currentKt = item.getModelObject();
                final String langCode = currentKt.getLangCode();
                final Iterator<KeywordTranslation> it = KeywordEditTranslationPanel.this
                    .getModelObject()
                    .getTranslations()
                    .get(langCode)
                    .iterator();
                while (it.hasNext()) {
                    final KeywordTranslation kt = it.next();
                    if (currentKt.equals(kt)) {
                        it.remove();
                        break;
                    }
                }
                target.add(getForm());
            }
        };
        newLink.add(new ButtonBehavior());
        newLink.setBody(new StringResourceModel("button." + id + LABEL_RESOURCE_TAG));
        newLink.add(new ConfirmationBehavior());
        return newLink;
    }

    protected abstract Form getForm();
}

File: core/core-web/src/main/java/ch/difty/scipamato/core/web/keyword/KeywordEditTranslationPanel.kt
package ch.difty.scipamato.core.web.keyword

import ch.difty.scipamato.common.web.AbstractPanel
import ch.difty.scipamato.core.entity.keyword.KeywordDefinition
import ch.difty.scipamato.core.entity.keyword.KeywordTranslation
import ch.difty.scipamato.core.web.common.DefinitionEditTranslationPanel
import de.agilecoders.wicket.core.markup.html.bootstrap.button.ButtonBehavior
import de.agilecoders.wicket.extensions.markup.html.bootstrap.confirmation.ConfirmationBehavior
import org.apache.wicket.ajax.AjaxRequestTarget
import org.apache.wicket.ajax.markup.html.AjaxLink
import org.apache.wicket.markup.html.basic.Label
import org.apache.wicket.markup.html.form.Form
import org.apache.wicket.markup.html.form.TextField
import org.apache.wicket.markup.repeater.Item
import org.apache.wicket.model.IModel
import org.apache.wicket.model.StringResourceModel

internal abstract class KeywordEditTranslationPanel(
    id: String,
    model: IModel<KeywordDefinition?>?
) : DefinitionEditTranslationPanel<KeywordDefinition?, KeywordTranslation>(id, model) {

    override fun addColumns(item: Item<KeywordTranslation>) {
        item.add(Label("langCode"))
        item.add(TextField<String>("name"))
        item.add(newAddLink("addTranslation", item))
        item.add(newRemoveLink("removeTranslation", item))
    }

    @Suppress("SameParameterValue")
    private fun newAddLink(id: String, item: Item<KeywordTranslation>): AjaxLink<Void> = object : AjaxLink<Void>(id) {
        override fun onClick(target: AjaxRequestTarget) {
            val currentKt = item.modelObject
            val langCode = currentKt.langCode
            val newKt = KeywordTranslation(null, langCode, null, 1)
            this@KeywordEditTranslationPanel.modelObject!!.addTranslation(langCode, newKt)
            target.add(form)
        }
    }.apply<AjaxLink<Void>> {
        add(ButtonBehavior())
        body = StringResourceModel("button." + id + AbstractPanel.LABEL_RESOURCE_TAG)
    }

    private fun newRemoveLink(id: String, item: Item<KeywordTranslation>): AjaxLink<Void> =
        object : AjaxLink<Void>(id) {
            override fun onClick(target: AjaxRequestTarget) {
                val currentKt = item.modelObject
                val keywordDefinition = this@KeywordEditTranslationPanel.modelObject!!
                keywordDefinition.removeTranslation(currentKt)
                target.add(form)
            }
        }.apply<AjaxLink<Void>> {
            add(ButtonBehavior())
            body = StringResourceModel("button." + id + AbstractPanel.LABEL_RESOURCE_TAG)
            add(ConfirmationBehavior())
        }

    protected abstract val form: Form<*>
}
